### PR TITLE
Automated Backup and Restoration System

### PR DESCRIPTION
**Pull request by jsmatias**
This backup system is designed to run within a Docker container using Cron jobs to schedule regular executions. 

**In summary, it includes:**

- **Dockerized Backup Solution**: Encapsulates the backup logic within a Docker container.

- **Cron Job Scheduling**: Utilises cron for scheduling backups at regular intervals. This approach provides flexibility in how frequently backups are performed, and for which service (MySQL, Keycloak, connectors, etc).

- **Incremental Backups**: It uses incremental backup logic implemented in cycles. Each cycle contains a full backup followed by incremental ones. Currently, it is set to run daily in weekly cycles.   

- **Restoring procedure**: The _restore.sh_ script is available to be run manually if any file needs to be restored.

### CODE REVIEW

File: src/tests/resources/connectors/zenodo/list_records_2.xml
Comment (jsmatias) [inline src/tests/resources/connectors/zenodo/list_records_2.xml:1]: These changes were made just to make vscode to stop complaining about not listed `resourceTypeGeneral`.

File: backups/backup.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+
+if [ $# -ne 3 ]; then
+  echo "Usage: $0 path/to/data path/to/backup/dir <cycle length:int>"
```
Comment (PGijsbers) [inline backups/backup.sh:4]: I think it would be good to give a little more documentation about cycle length.

File: src/tests/backups/test_backup_scripts.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        os.makedirs(original_dir)
+        os.makedirs(backup_dir)
+        create_test_files(original_dir, ["file1.txt", "file2.txt"])
+
+        cycle_length = 2
+        backup_cycle_0_path = backup_dir / "original" / "original_0"
+        backup_cycle_1_path = backup_dir / "original" / "original_1"
+        files_in_cycle_0 = [
+            backup_cycle_0_path / "original0.tar.gz",
+            backup_cycle_0_path / "original1.tar.gz",
```
Comment (PGijsbers) [inline src/tests/backups/test_backup_scripts.py:45]: I assume this would be a file of insignificant size as there is no delta between the first and second backup?

File: backups/backup.sh (hunk 1 above)
Comment (PGijsbers) [reply backups/backup.sh:4]: and a link to e.g., https://www.gnu.org/software/tar/manual/html_section/Incremental-Dumps.html

File: src/tests/backups/test_backup_scripts.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            backup_cycle_0_path / "original.snar",
```
Comment (PGijsbers) [inline src/tests/backups/test_backup_scripts.py:46]: It looks like you can delete this file after completing a full cycle? Though keeping it around probably doesn't hurt as it should only be meta-data. Keeping it also reduces the complexity of the script and potentially allows for continued backups at a higher level.

**Review by PGijsbers — CHANGES_REQUESTED** [review]
Some minor suggestions. Mainly regarding documentation of the cycle functionality. The implementation seems OK. Did not try to run this yet.

File: backups/entry.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+mkdir -p /opt/backups/data/
+
+# Run cron on the foreground with log level WARN
+/usr/sbin/cron -f -l 4
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply backups/entry.sh:6]: That sounds good, I'll do it.

File: backups/backup.sh [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+backup_file="${backup_dir}/${data_to_backup}${level}.tar.gz"
+snapshot_file="${backup_dir}/${data_to_backup}.snar"
+
+echo $backup_file
+echo $snapshot_file
+
+tar -czf "$backup_file" --listed-incremental="$snapshot_file" "$path_to_backup"
+
+echo $(date -u) "$data_to_backup done!"
+
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply backups/backup.sh:73]: I agree. It just got longer than I anticipated. In principle it should be just a simple `tar` command plus a few conditions. When I realised it was already around 100 lines. I'll pay better attention for future ones.

File: src/tests/backups/test_backup_scripts.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        ]
+
+        call_backup_command(backup_command)
+        os.remove(original_dir / "file1.txt")
+        with open(original_dir / "file2.txt", "w") as f:
+            f.write("Content of new file\n")
+        call_backup_command(backup_command)
+
+        call_restore_command(restore_command)
+        assert not os.path.exists(restored_file1_path), f"{restored_file1_path} shouldn't be here."
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply src/tests/backups/test_backup_scripts.py:109]: Well noticed! Thanks, I replaced it.

File: src/tests/backups/test_backup_scripts.py (hunk 3 above)
Comment (jsmatias) [reply src/tests/backups/test_backup_scripts.py:46]: Yes, you're correct. For now, I believe it's fine to keep old backups just in case, since we're not yet assessing or validating the metadata quality. In future, it will be better to implement some sort of deletion of older backups, then we would keep just one or a few ones.

File: src/tests/backups/test_backup_scripts.py (hunk 2 above)
Comment (jsmatias) [reply src/tests/backups/test_backup_scripts.py:45]: Yes, you're correct. If no change was made to the original files after the backup of level 0 (_original0.tar.gz_), then the next level of backup (_original1.tar.gz_) will be empty. This is information of the state of the last backup is stored in the _.snar_ files.

File: backups/backup.sh (hunk 1 above)
Comment (jsmatias) [reply backups/backup.sh:4]: Sure, thanks. 
I added detailed information about it for both scripts, _backup.sh_ and _restore.sh_.

**Comment by jsmatias** [comment]
After addressing most of the feedback from previous comments, I encountered an issue where the unit tests for this PR ceased to function correctly, even when using the same commit tag that had previously passed. Furthermore, using a `subprocess` in Python to execute the `tar` command for pytests proved to be suboptimal. This approach introduced a dependency on the operating system hosting the tests, specifically due to the requirement of having GNU tar for this backup service to function correctly.

Considering the critical role of these tests, I've relocated their execution to occur inside the backup container itself, initiated from the entry point. This adjustment ensures that the backup container will only start backup operations once it has successfully passed these tests.

To circumvent the dependency on the host OS having GNU tar installed for the restoration script, I've added an `_entry_restore.sh` script. This script is designed to be executed within the container whenever a backup restoration is needed.

In summary, the following modifications have been implemented:
- Unit tests have been moved to execute from within the container, initiated at the entry point.
- The logic for data restoration has been integrated into the container environment.
- A `readme.md` file has been added to enhance the documentation, providing clear instructions on how to perform backups and restorations.

File: backups/backup.sh (hunk 1 above)
Comment (jsmatias) [reply backups/backup.sh]: I also added a README.md to describe the process

File: scripts/mysql_restore.sh [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+cd $SCRIPT_DIR/..
+source .env
+
+DATA_PATH=$(realpath "$DATA_PATH")
+
+echo $(date -u) "Copying backup.sql to container..."
+docker cp "${DATA_PATH}/mysql_dump/backup.sql" sqlserver:/tmp/backup.sql
+
+echo $(date -u) "Restoring database..."
+docker exec -i /bin/bash -c "mysql -uroot -p${MYSQL_ROOT_PASSWORD} aiod < /tmp/backup.sql"
```
Comment (PGijsbers) [inline scripts/mysql_restore.sh:15]:
```suggestion
docker exec -i sqlserver /bin/bash -c "mysql -uroot -p${MYSQL_ROOT_PASSWORD} aiod < /tmp/backup.sql"
```

**Review by PGijsbers — CHANGES_REQUESTED** [review]
Hi,

I left some feedback. Overall it looks good. Unfortunately, I couldn't test this that well, it seems that `tar` on macos does not do incremental backups :/ And this does worry me, since it seems that you didn't run this either? Because a critical part of the `mysql_restore` call is missing (the docker container name).

I can test this in a linux environment later, but figured I will send this feedback already.

**Comment by jsmatias** [comment]
> Hi,
> 
> I left some feedback. Overall it looks good. Unfortunately, I couldn't test this that well, it seems that `tar` on macos does not do incremental backups :/ And this does worry me, since it seems that you didn't run this either? Because a critical part of the `mysql_restore` call is missing (the docker container name).
> 
> I can test this in a linux environment later, but figured I will send this feedback already.

Indeed, `tar` on MacOS doesn't have incremental backups. But I installed `GNU tar` to to run it locally. 
In fact, the docker container name was missing! Thanks for spotting that. 
What happened is that when I was testing it, I ran the command just once directly from terminal and when I passed the commands to the script I forgot the container's name. I ended up not testing it again, because the process to restore the entire DB takes some time. 

I cleaned up the DB to make it quicker and tested it properly this time.

File: scripts/mysql_restore.sh (hunk 7 above)
Comment (jsmatias) [reply scripts/mysql_restore.sh:15]: Applied the change in commit a2a6741.

File: scripts/README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+
+### `database-connect.sql`
+
+- **Purpose**: Connects to the SQL server. This SQL script should be run after executing `docker-compose up`, ensuring that the SQL server is already running.
+- **Usage**: Execute this SQL script against your SQL server instance to establish a connection.
+
+### `mysql_dump.sh`
+
+- **Purpose**: Exports a logical backup of the `aiod` database from the SQL server and extracts it to `data/path/mysql_dump`.
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply scripts/README.md:20]: Did you mean to add some change here?

File: scripts/README.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+## Script Descriptions
+
+### `clean.sh`
+
+**Warning!** - This deletes all the content in the data/ directory!
+- **Purpose**: This convenience script reverts the development environment back to a clean state, including resetting the database.
+- **Usage**: Simply run the script without any arguments.
+
+
+### `database-connect.sql`
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply scripts/README.md:14]: Good point. I updated the file name as well.

File: scripts/README.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
+# Scripts README.md
+
+This directory contains a collection of scripts designed to manage backups, restorations, and database interactions for the service. Below is a detailed description of each script and its intended usage.
+
+## Script Descriptions
+
+### `clean.sh`
+
+**Warning!** - This deletes all the content in the data/ directory!
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply scripts/README.md:10]: Cool! I added it.

File: scripts/realm_export.sh [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+DATA_PATH=$(realpath "$DATA_PATH")
+LOCAL_BACKUP_PATH="$DATA_PATH"/keycloak_realm
+
+docker exec -i keycloak /bin/bash -c "/opt/keycloak/bin/kc.sh export --file /tmp/aiod.json --realm aiod"
+
+if [ ! -d "$LOCAL_BACKUP_PATH" ]; then
+    mkdir "$LOCAL_BACKUP_PATH"
+fi
+
+docker cp keycloak:/tmp/aiod.json "$LOCAL_BACKUP_PATH"/aiod.json
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply scripts/realm_export.sh:18]: Sure, we can do that. 👍

File: scripts/mysql_dump.sh [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+
+DATA_PATH=$(realpath "$DATA_PATH")
+LOCAL_BACKUP_PATH="$DATA_PATH"/mysql_dump
+
+docker exec -i sqlserver /bin/bash -c "
+    mysqldump -uroot -p"$MYSQL_ROOT_PASSWORD" aiod > /tmp/backup.sql
+"
+
+if [ ! -d "$LOCAL_BACKUP_PATH" ]; then
+    mkdir "$LOCAL_BACKUP_PATH"
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply scripts/mysql_dump.sh:17]: It's fine. This path should be dedicated to the backups only. If a file with same name is there, this should be addressed before proceeding with the backups. This way any conflict can be avoided.

File: scripts/realm_export.sh [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+cd $SCRIPT_DIR/..
+
+source .env
+DATA_PATH=$(realpath "$DATA_PATH")
+LOCAL_BACKUP_PATH="$DATA_PATH"/keycloak_realm
+
+docker exec -i keycloak /bin/bash -c "/opt/keycloak/bin/kc.sh export --file /tmp/aiod.json --realm aiod"
+
+if [ ! -d "$LOCAL_BACKUP_PATH" ]; then
+    mkdir "$LOCAL_BACKUP_PATH"
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply scripts/realm_export.sh:15]: It's okay, same as above.

File: scripts/restore.sh [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+done
+
+echo "-> Initiating..."
+for ((i = 0; i <= level; i++)); do
+    backup_file="${data_to_restore}${i}.tar.gz"
+    backup_file_path="${backup_dir}/${backup_file}"
+    echo "   Restoring: ${backup_file}"
+    tar --directory="$destination_dir" --extract --file="$backup_file_path" --listed-incremental=/dev/null
+done
+;;
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply scripts/restore.sh]: Indeed. This was a bit messy, I moved it out and replaced it by an `if` statement in the commit [18449b3](https://github.com/aiondemand/AIOD-rest-api/pull/275/commits/18449b3b65707d0b21abd5637f0cce5fe5feb813).

File: src/tests/backups/test_backup_restore.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+def call_backup_command(command: list[str], env: dict) -> subprocess.CompletedProcess:
+    return subprocess.run(command, env=env, check=True)
+
+
+def call_restore_command(command: list[str], env: dict) -> subprocess.CompletedProcess:
+    return subprocess.run(command, env=env, input="y\n", text=True, check=True)
+
+
+def test_backup_happy_path():
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply src/tests/backups/test_backup_restore.py:47]: Yes, the scripts require GNU tar. I added a condition to check if GNU tar is installed and available in the system, otherwise the tests are skipped.

File: scripts/restore.sh [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    backup_file="${backup_dir}/${data_to_restore}${i}.tar.gz"
+    if [ $i -le $level ] || [ $i -eq 0 ]; then
+        check_file_or_dir "$backup_file"
+    fi
+    if [ $i -eq $((level + 1)) ] || [ ! -f "$backup_file" ]; then
+        concluded=true
+        level=$((i - 1))
+    fi
+    i=$((i + 1))
+done
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply scripts/restore.sh]: I agree, this is not very intuitive. I added some comments to the file in commit [3b144ba](https://github.com/aiondemand/AIOD-rest-api/pull/275/commits/3b144bae6543afe1a8c1a80c893c344e9871ec12).

File: scripts/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (PGijsbers) [reply scripts/README.md:21]: I am not sure (anymore) what happened here. Please ignore - likely an artifact from me restructuring some suggestions half-way through the document.

File: scripts/backup.sh [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        level=0
+    elif [ "$level" -ge "$cycle_length" ]; then
+        echo $(date -u) "-> Backup cycle completed, starting a new one."
+        last_backup_label="${backup_dir##*_}"
+        backup_dir="${parent_dir}/${data_to_backup}_$((last_backup_label + 1))"
+        mkdir "$backup_dir"
+        level=0
+    fi
+fi
+
```
Comment (PGijsbers) [inline scripts/backup.sh:67]: Printing some information on which cycle and level are used will be useful information to have in the logs.

File: scripts/restore.sh [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+
+echo "-> Verifying paths..."
+check_file_or_dir $backup_dir
+check_file_or_dir $destination_dir
+
+echo ""
+echo "Are you sure you want to proceed? (y/n)"
+echo "This overwrites all files in the destination directory and deletes the ones which are not in the archive!"
+echo ""
+echo "$backup_dir -------> $destination_dir/$data_to_restore"
```
Comment (PGijsbers) [inline scripts/restore.sh:74]:
```suggestion
echo ""
echo "$backup_dir -------> $destination_dir/$data_to_restore"
echo ""
echo "This overwrites all files in the destination directory and deletes the ones which are not in the archive!"
echo "Are you sure you want to proceed? (y/n)"
```

The prompt is easier to miss if it's not the last thing on the screen.

**Review by PGijsbers — APPROVED** [review]
Got around to testing it (sorry for the delay). Seems to work as expected/described (tested only with the database backup, adding/removing rows). Left minor comments from actually using it, but feel free to merge without a new review round.

File: scripts/restore.sh (hunk 18 above)
Comment (jsmatias) [reply scripts/restore.sh:74]: Nice! Good point.
